File: lab4/ex03/ex0402b/src/turtle2_tf2_broadcaster.cpp
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <tf2_ros/transform_broadcaster.h>
#include <turtlesim/msg/pose.hpp>
#include "ex0402b/utilities.hpp"

class Turtle2Tf2Broadcaster : public rclcpp::Node
{
public:
  Turtle2Tf2Broadcaster()
  : Node("turtle2_tf2_broadcaster")
  {
    tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
    
    subscription_ = this->create_subscription<turtlesim::msg::Pose>(
      "/turtle2/pose", 1,
      std::bind(&Turtle2Tf2Broadcaster::handle_turtle_pose, this, std::placeholders::_1));
    
    RCLCPP_INFO(this->get_logger(), "Started turtle2_tf2_broadcaster");
  }

private:
  void handle_turtle_pose(const std::shared_ptr<turtlesim::msg::Pose> msg)
  {
    geometry_msgs::msg::TransformStamped t;
    
    t.header.stamp = this->get_clock()->now();
    t.header.frame_id = "world";
    t.child_frame_id = "turtle2";
    
    t.transform.translation.x = msg->x;
    t.transform.translation.y = msg->y;
    t.transform.translation.z = 0.0;
    
    auto q = ex0402b::quaternion_from_euler(0, 0, msg->theta);
    t.transform.rotation = q;
    
    tf_broadcaster_->sendTransform(t);
  }

  std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<Turtle2Tf2Broadcaster>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: lab4/ex02/ex0402a/ex0402a/turtle2_tf2_broadcaster.py
import math
from geometry_msgs.msg import TransformStamped
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from turtlesim.msg import Pose

class Turtle2Tf2Broadcaster(Node):

    def __init__(self):
        super().__init__('turtle2_tf2_broadcaster')
        
        # Initializing the transform broadcaster:
        self.tf_broadcaster = TransformBroadcaster(self)
        
        # Subscribing to turtle2 pose:
        self.subscription = self.create_subscription(
            Pose,
            '/turtle2/pose',
            self.handle_turtle_pose,
            1)
        
        self.get_logger().info('Started turtle2_tf2_broadcaster')

    def handle_turtle_pose(self, msg):
        t = TransformStamped()
        
        # Reading message content and assign it to corresponding tf variables:
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'world'
        t.child_frame_id = 'turtle2'
        
        # Turtle only exists in 2D, thus we get x and y translation
        # coordinates from the message and set the z coordinate to 0:
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0
        
        # For the same reason, turtle can only rotate around one axis
        # and this why we set rotation in quaternion:
        q = self.quaternion_from_euler(0, 0, msg.theta)
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]
        
        # Sending the transformation:
        self.tf_broadcaster.sendTransform(t)

    def quaternion_from_euler(self, roll, pitch, yaw):
        cy = math.cos(yaw * 0.5)
        sy = math.sin(yaw * 0.5)
        cp = math.cos(pitch * 0.5)
        sp = math.sin(pitch * 0.5)
        cr = math.cos(roll * 0.5)
        sr = math.sin(roll * 0.5)
        
        q = [0] * 4
        q[0] = sr * cp * cy - cr * sp * sy
        q[1] = cr * sp * cy + sr * cp * sy
        q[2] = cr * cp * sy - sr * sp * cy
        q[3] = cr * cp * cy + sr * sp * sy
        
        return q

def main():
    rclpy.init()
    node = Turtle2Tf2Broadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
        
    rclpy.shutdown()